hyperspy/model.py

```diff
@@ -806,18 +830,42 @@ def fit(self, fitter=None, method='ls', grad=False,
 
             # Least squares "dedicated" fitters
             if fitter == "leastsq":
-                output = \
-                    leastsq(self._errfunc, self.p0[:], Dfun=jacobian,
-                            col_deriv=1, args=args, full_output=True, **kwargs)
-
-                self.p0, pcov = output[0:2]
-                signal_len = sum([axis.size
-                                  for axis in self.axes_manager.signal_axes])
-                if (signal_len > len(self.p0)) and pcov is not None:
-                    pcov *= ((self._errfunc(self.p0, *args) ** 2).sum() /
-                             (len(args[0]) - len(self.p0)))
+                if bounded:
+                    self.set_boundaries()
+                    ls_b = self.free_parameters_boundaries
+                    ls_b = ([ a if a is not None else -np.inf for a,b in ls_b ],
+                            [ b if b is not None else np.inf for a,b in ls_b ])
+                    output = \
+                        least_squares(self._errfunc, self.p0[:],
+                                      args=args, bounds=ls_b, **kwargs)
+
+                    # Do Moore-Penrose inverse discarding zero singular values
```

Comment:
This code was actually lifted from `curve_fit` - but yes I can switch to `pinv`. With regards to `curve_fit`, it takes `curve_fit(f, xdata, ydata, ...)` so I found it was actually easier to call least_squares directly.